src/IO.Ably.Shared/Realtime/Presence.cs

```
@@ -312,79 +467,115 @@ public bool Put(PresenceMessage item)
                 lock (_lock)
                 {
                     // we've seen this member, so do not remove it at the end of sync
-                    residualMembers?.Remove(item.MemberKey);
+                    _residualMembers?.Remove(item.MemberKey);
                 }
 
-                // compare the timestamp of the new item with any existing member (or ABSENT witness)
-                PresenceMessage existingItem;
-                if (members.TryGetValue(item.MemberKey, out existingItem) && (item.Timestamp < existingItem.Timestamp))
-                    return false;
+                try
+                {
+                    if (_members.TryGetValue(item.MemberKey, out var existingItem) && existingItem.IsNewerThan(item))
+                    {
+                        return false;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    if (Logger.IsDebug) Logger.Debug(ex.Message);
```

Should this not be error level at least given this shouldn't happen right?  As far as I am concerned, a failure like this should probably even result in the channel becoming detached because we have an unknown fault and as such can no longer guarantee the state of the channel is correct.  Do you not think this is a channel failure?

---

I didn't actually add this code, it comes from the original PR. 
That said, I am not convinced there should be a try/catch block here at all, the use of TryGetValue means only place where an exception can bubble up here is from the call to existingItem.IsNewerThan which parses the presence message id.

---

Well I think if there is a risk that we can crash people's apps, then it would make sense to catch it, change state to failed, and make sure we log that error.  Generally that would be my preference.

---

I think the approach in the [Java version](https://github.com/ably/ably-java/blob/159018c30b3ef813a9d3ca3c6bc82f51aacbbc68/lib/src/main/java/io/ably/lib/realtime/Presence.java#L901) make sense, the responsibility is placed with `hasNewerItem` method. The [calling code](https://github.com/ably/ably-java/blob/159018c30b3ef813a9d3ca3c6bc82f51aacbbc68/lib/src/main/java/io/ably/lib/realtime/Presence.java#L853) has no exception handling. I do think we should add some more test coverage to check the handling of malformed id strings though.

---

@withakay I spoke with @SimonWoolf and we both agree that the Java implementation is certainly wrong.  Silently failing when doing a compare means we have a malformed message and thus the entire protocol at that point is compromised.  The only thing at that point we should do is raise an exception, log the problem, and move the channel to the failed state.  I am going to raise an issue in Java to correct this.   Silently catching exceptions in almost all cases is wrong.

---

See https://github.com/ably/ably-java/issues/357

---

@mattheworiordan OK, understood. At that point in the code the exception would be in the PresenceMap, so I will probably want move the exception handling up into the Presence class in order to access the Channel to move it into a failed state.

---

Yup, makes sense, I'll trust you to work out how to propagate up! 👍